portal/migrations/versions/55469bdd181f_.py

```diff
@@ -0,0 +1,111 @@
+from portal.models.user import User, active_patients
+from portal.models.overall_status import OverallStatus
+
+# revision identifiers, used by Alembic.
+revision = '55469bdd181f'
+down_revision = '9026af5fe040'
+
+
+def qb_pairs_by_rp():
+    """Lookup and generate a map of RP2->RP3 versions of QBs
+
+    Issue only applies to IRONMAN RPs - ignore others
+    """
+    rp2_id = ResearchProtocol.query.filter_by(name='IRONMAN v2').one().id
+    rp3_id = ResearchProtocol.query.filter_by(name='IRONMAN v3').one().id
+
+    qb_map = dict()  # keyed by the RP2 QB ID
+    for qb in QuestionnaireBank.query.filter_by(research_protocol_id=rp2_id):
+        rp3_name = "IRONMAN_v3" + qb.name[len('IRONMAN'):]
+        rp3_qb = QuestionnaireBank.query.filter_by(name=rp3_name).one()
+        assert (rp3_qb.research_protocol_id == rp3_id)
+        qb_map[qb.id] = rp3_qb.id
+    return qb_map
+
+
+def upgrade():
+    # Tricky task to locate the errors.  One time expensive process,
+    # loop through patients identifying problems.
+
+    qb_map = qb_pairs_by_rp()
+    fix_map = dict()
+
+    for patient in active_patients(include_test_role=True).order_by(User.id):
+        qnrs = QuestionnaireResponse.query.filter(
+            QuestionnaireResponse.subject_id == patient.id).order_by(
+            QuestionnaireResponse.authored)
+        if not qnrs.count():
+            # Only care if the user submitted QNRs:
+            continue
+
+        for qnr in qnrs:
+
+            # Obtain qb_timeline row at the time of QNR.authored
+            as_of_date = qnr.authored
+            qbt = QBT.query.filter(QBT.user_id == patient.id).filter(
+                QBT.at <= as_of_date).order_by(
+                QBT.at.desc(), QBT.id.desc()).first()
+
+            if qnr.questionnaire_bank_id is None:
+                print ("user_id {} authored {} {} with NO qb_id set".format(
+                    patient.id,
+                    as_of_date,
+                    qnr.document['questionnaire']['reference']))
+                continue
+            if qbt and qbt.status != OverallStatus.withdrawn:
+                recorded_qb = QuestionnaireBank.query.get(
+                    qnr.questionnaire_bank_id)
+                qbd = qbt.qbd()
+                expected_qb = qbd.questionnaire_bank
+                if recorded_qb.classification == 'indefinite':
+                    # indefinite case is handled differently...
+                    qb_status = QB_Status(patient, as_of_date)
+                    expected_qb = QuestionnaireBank.query.get(
+                        qb_status._current_indef.qb_id)
+
+                if recorded_qb != expected_qb:
+                    replacement = qb_map.get(recorded_qb.id)
+                    if replacement == expected_qb.id:
+                        # retain the change request, as altering it inline
+                        # interrupts the iteration
+                        fix_map[qnr.id] = expected_qb.id
+                        print(
```

Comment:
Could you make this an audit log entry instead?

Reply:
funny you should ask!  i had the same thought - see latest commit.